internal/ignition/ignition.go

```diff
@@ -0,0 +1,378 @@
+	bootstrapPath := filepath.Join(g.workDir, "bootstrap.ign")
+	err = g.updateBootstrap(bootstrapPath)
+	if err != nil {
+		return err
+	}
+
+	// move all files into the working directory
+	err = os.Rename(filepath.Join(g.workDir, "auth/kubeadmin-password"), filepath.Join(g.workDir, "kubeadmin-password"))
+	if err != nil {
+		return err
+	}
+	// after installation completes, a new kubeconfig will be created and made
+	// available that includes ingress details, so we rename this one
+	err = os.Rename(filepath.Join(g.workDir, "auth/kubeconfig"), filepath.Join(g.workDir, "kubeconfig-noingress"))
+	if err != nil {
+		return err
+	}
+	err = os.Remove(filepath.Join(g.workDir, "auth"))
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func bmhIsMaster(bmh *bmh_v1alpha1.BareMetalHost) bool {
+	return strings.Contains(bmh.Name, "-master-")
+}
+
+// updateBootstrap adds a status annotation to each BareMetalHost defined in the
+// bootstrap ignition file
+func (g *installerGenerator) updateBootstrap(bootstrapPath string) error {
+	bootstrapBytes, err := ioutil.ReadFile(bootstrapPath)
+	if err != nil {
+		g.log.Errorf("error reading file %s: %v", bootstrapPath, err)
+		return err
+	}
+
+	config, _, err := config_31.Parse(bootstrapBytes)
+	if err != nil {
+		g.log.Errorf("error parsing bootstrap ignition: %v", err)
+		return err
+	}
+
+	masters, workers := sortHosts(g.cluster.Hosts)
+	for i, file := range config.Storage.Files {
+		// workaround for https://github.com/openshift/machine-config-operator/issues/2086
+		if isMOTD(&config.Storage.Files[i]) {
+			g.fixMOTDFile(&config.Storage.Files[i])
+		}
+		if isBMHFile(&config.Storage.Files[i]) {
+			// extract bmh
+			bmh, err := fileToBMH(&config.Storage.Files[i]) //nolint,shadow
+			if err != nil {
+				g.log.Errorf("error parsing File contents to BareMetalHost: %v", err)
+				return err
+			}
+
+			// get corresponding host
+			var host *models.Host
+			if bmhIsMaster(bmh) {
+				if len(masters) == 0 {
+					return fmt.Errorf("Not enough registered masters to match with BareMetalHosts")
+				}
+				host, masters = masters[0], masters[1:]
+			} else {
+				if len(workers) == 0 {
+					return fmt.Errorf("Not enough registered workers to match with BareMetalHosts")
+				}
+				host, workers = workers[0], workers[1:]
+			}
+
+			// modify bmh
+			g.log.Infof("modifying BareMetalHost ignition file %s", file.Node.Path)
+			err = g.modifyBMHFile(&config.Storage.Files[i], bmh, host)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	// write ignition back to disk
+	updatedBytes, err := json.Marshal(config)
+	if err != nil {
+		return err
+	}
+
+	err = ioutil.WriteFile(bootstrapPath, updatedBytes, 0600)
+	if err != nil {
+		g.log.Errorf("error writing file %s: %v", bootstrapPath, err)
+		return err
+	}
+	g.log.Infof("Updated file %s", bootstrapPath)
+
+	return nil
+}
+
+func isBMHFile(file *config_31_types.File) bool {
+	return strings.Contains(file.Node.Path, "openshift-cluster-api_hosts")
```

Comment:
@yevgeny-shnaidman is there a better way?

Reply:
FWIW this is a straight port of how the python implementation works.